Clean up build_info.h after 1.0

gilles-peskine-arm opened the pull request:
Clean up some things in `build_info.h` and the header that it includes:

* Resolves https://github.com/Mbed-TLS/mbedtls/issues/10305.
* Move config adjustment headers to `.../private` — this had only been partially done in 1.0.
* Sort out which config adjustments belong under `/include` and which ones belong under `/drivers/builtin/include`.
* Rename config adjustment headers to follow a more systematic naming scheme, and to better reflect their role in a world where the user configuration is only driven by PSA.

Needs preceding PR: framework https://github.com/Mbed-TLS/mbedtls-framework/pull/236

I used (and even wrote) https://github.com/Mbed-TLS/mbedtls-docs/pull/193 to check for mistakes when reordering parts of `build_info.h`.

## PR checklist

- [x] **changelog** provided | not required because: only internal changes
- [x] **framework PR** provided https://github.com/Mbed-TLS/mbedtls-framework/pull/236
- [x] **mbedtls development PR** not required because: only internal changes in crypto
- [x] **mbedtls 3.6 PR** not required because: new stuff
- **tests**  provided

gilles-peskine-arm committed:
Remove obsolete inclusion of limits.h

Individual headers or C files are supposed to include what they need.

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

gilles-peskine-arm committed:
Move stray dependency inclusion where it belongs

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

gilles-peskine-arm committed:
Move crypto_adjust_config_synonyms.h where it belongs

This is at the PSA level, not driver-specific, so keep it in `/include`.

This is not meant to be included directly by applications, so move it under
`.../private`.

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

gilles-peskine-arm committed:
Move crypto_adjust_config_dependencies where it belongs

This is changing PSA configurations, based on conditions that are not
driver-specific, so keep it in `/include`. The file has guards based on
whether some key derivation algorithms are accelerated, but these guards are
currently irrelevant since key derivation drivers are not supported yet.

This is not meant to be included directly by applications, so move it under
`.../private`.

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

gilles-peskine-arm committed:
Move crypto_adjust_config_key_pair_types where it belongs

This is at the PSA level, not driver-specific, so keep it in `/include`.

This is not meant to be included directly by applications, so move it under
`.../private`.

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

gilles-peskine-arm committed:
Move crypto_adjust_config_auto_enabled.h where it belongs

This commit changes the location of the file in the inclusion order, to
group it with other headers that tweak the PSA configuration. The former
order also worked because the macros that it references aren't used in
other configuration-tweaking headers at this time (and have never been,
I believe), so this is not a bug fix, just a readability and robustness
improvement.

This is at the PSA level, not driver-specific, so keep it in `/include`.

This is not meant to be included directly by applications, so move it under
`.../private`.

Rename the file to match the pattern followed by similar files (I had
forgotten a word when I created that file initially).

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

gilles-peskine-arm committed:
Move crypto_adjust_config_derived.h where it belongs

This commit changes the location of the file in the inclusion order, to
group it with other headers that tweak the PSA configuration. Specificaly,
this header is now included before `config_adjust_legacy_from_psa.h` and
`config_adjust_test_accelerators.h`. Both of these set driver-specific
macros based on `PSA_WANT` macros, whereas
`crypto_adjust_config_derived.h` does not query driver-specific macros, so
the new order is no worse.

In fact the new order more correct because it would make sense for
`config_adjust_legacy_from_psa.h` to query macros set by
`crypto_adjust_config_derived.h`. However, it currently doesn't, so this
is not a bug fix, just a robustness improvement.

This is at the PSA level, not driver-specific, so keep it in `/include`.

This is not meant to be included directly by applications, so move it under
`.../private`.

Rename the file to match the pattern followed by similar files (I had
forgotten a word when I created that file initially).

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

gilles-peskine-arm committed:
Meld config_psa.h into build_info.h

Historically, we did config ajustments in a separate header `config_psa.h`.
Then we split `config_psa.h` into pieces. Today, having `config_psa.h` as a
layer of indirection in header inclusions is no longer useful. So remove it.

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

gilles-peskine-arm committed:
Update name of crypto_adjust_config_enable_builtins.h

Rename the file to match the pattern followed by similar files, and to
better reflect the role of the file in the context of TF-PSA-Crypto. It's no
longer about enabling legacy interfaces, since low-level crypto is mostly no
longer part of a public interface, but about enabling internal interfaces
that implement actual cryptography.

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

gilles-peskine-arm committed:
Move config_adjust_legacy_crypto.h where it belongs

This is specific to modules that are under `/drivers`, so keep it under
`/drivers`. Some of these modules are also public interfaces that are used
in builds with only third-party drivers, so the corresponding macros should
be outside of `/drivers`; this will be handled in a subsequent commit.

This is not meant to be included directly by applications, so move it under
`.../private`.

Rename the file to match the pattern followed by similar files, and to
better reflect its current role: it's tweaking modules that implement the
built-in cryptography, which were formerly legacy public interfaces but are
now internal.

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

gilles-peskine-arm committed:
Create a new header file for config adjustments for support modules

This one comes after both the PSA feature set and the feature set of
built-in crypto have been determined.

This commit creates an empty file. It will be populated in a subsequent commit.

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

gilles-peskine-arm committed:
Move parts of crypto_adjust_config_tweak_builtins.h where they belong

Split out the parts of `crypto_adjust_config_tweak_builtins.h` that are
actually about generic support modules (platform, ASN.1, MD, ...), and
move them outside `/drivers`, into `crypto_adjust_config_support.h`.

One macro only depends on PSA options, so move it to
`crypto_adjust_config_derived.h` instead.

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

gilles-peskine-arm committed:
Update framework supporting new location of crypto_adjust_config_synonyms.h

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

gilles-peskine-arm committed:
Be more robust about include directories

Install all headers without hard-coding which sub(sub)directories exist.

Don't use file(GLOB), following CMake best practices (the list might not be
updated if a source file is added or removed when the build tree already
exists). It isn't actually useful anyway.

To keep things simple, don't permissions: we just want the default permissions.

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

gilles-peskine-arm committed:
Define MBEDTLS_MD_LIGHT before checking it

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

ronald-cron-arm reviewed (changes requested):
This looks almost good to me. I have only a few comments on the last commit.
Sorry, I was aiming to post that on the associated framework PR.

ronald-cron-arm commented on include/tf-psa-crypto/private/crypto_adjust_config_dependencies.h:
```diff
@@ -42,14 +43,16 @@
 #define PSA_WANT_KEY_TYPE_HMAC 1
 #endif
 
+/* Generic implementation o some key derivation algorithms based on CMAC */
```
```suggestion
/* Generic implementation of some key derivation algorithms based on CMAC */
```

ronald-cron-arm commented on include/tf-psa-crypto/private/crypto_adjust_config_derived.h:
```diff
@@ -1,5 +1,5 @@
 /**
- * \file psa/crypto_adjust_config_derived.h
+ * \file tf-psa-crypto/private/crypto_adjust_config_derived.h
```
There is a reference to "psa/crypto_adjust_config_derived.h" to update in `entropy.c`.

ronald-cron-arm commented on include/tf-psa-crypto/build_info.h:
```diff
@@ -143,23 +143,36 @@
+
+/* Special header to adjust the configuration to make a build
+ * where all enabled mechanisms are provided both as built-in and
+ * through libtestdriver1. See the comment at the top of the
```
```suggestion
 * through test drivers. See the comment at the top of the
```
This is not used in `libtestdriver1` context I think.

ronald-cron-arm commented on drivers/builtin/include/mbedtls/private/crypto_adjust_config_enable_builtins.h:
```diff
@@ -1,21 +1,21 @@
 /**
- * \file mbedtls/config_adjust_legacy_from_psa.h
- * \brief Adjust PSA configuration: activate legacy implementations
+ * \file mbedtls/private/crypto_adjust_config_enable_builtins.h
```
`config_adjust_legacy_from_psa.h` is still referenced in `config_adjust_test_accelerators.h`.

ronald-cron-arm commented on drivers/builtin/include/mbedtls/private/crypto_adjust_config_tweak_builtins.h:
```diff
@@ -1,28 +1,21 @@
 /**
- * \file mbedtls/config_adjust_legacy_crypto.h
- * \brief Adjust legacy configuration configuration
+ * \file mbedtls/private/crypto_adjust_config_tweak_builtins.h
```
config_adjust_legacy_crypto.h is still referenced in:
`include/tf-psa-crypto/private/crypto_adjust_config_derived.h`
`core/tf_psa_crypto_check_config.h`
`drivers/builtin/include/mbedtls/private/config_adjust_test_accelerators.h`

ronald-cron-arm commented on include/tf-psa-crypto/private/crypto_adjust_config_support.h:
```diff
@@ -0,0 +1,219 @@
+#define TF_PSA_CRYPTO_PRIVATE_CRYPTO_ADJUST_CONFIG_SUPPORT_H
+
+#if !defined(TF_PSA_CRYPTO_CONFIG_FILES_READ)
+#error "Do not include mbedtls/config_adjust_*.h manually! This can lead to problems, " \
```
The "Do not include mbedtls/config_adjust_*.h" part should be updated.
```suggestion
#error "Do not include crypto_adjust_config_*.h manually! This can lead to problems, " \
```
?

gilles-peskine-arm replied:
Actually I think this error is obsolete, now that the header is under `private`. We added this error because there used to be no indication that the header wasn't a public interface. So I'll remove it.

ronald-cron-arm commented on include/tf-psa-crypto/private/crypto_adjust_config_support.h:
It would probably be better to update that.

gilles-peskine-arm replied:
To what? We could split `adjust_config_support.h` further into `adjust_config_platform.h` and `adjust_config_extra.h`, but I don't think it would help materially at this point — mainly I wanted to split between `adjust_config_support.h` under `/include` and `adjust_config_tweak_builtins.h` under `drivers/builtin/include`.

ronald-cron-arm replied:
My comment was not clear. I was referring to the references to "mbedtls_config.h" lines 27 and 28. This should be crypto_config.h now I guess.
Related to that the comment about MBEDTLS_PLATFORM_SNPRINTF_ALT in crypto_config.h is outdated:
```
 * \note MBEDTLS_PLATFORM_SNPRINTF_ALT is required on Windows;                  
 * it will be enabled automatically by check_config.h
```

ronald-cron-arm commented on include/tf-psa-crypto/private/crypto_adjust_config_support.h:
This may be too late in the inclusion order. I can see some "#if defined(MBEDTLS_MD_C)" above.

ronald-cron-arm reviewed (changes requested):
Thanks for this nice clean-up. Thanks for the commit message that helped a lot reviewing the changes. I have some minor comments/questions.

gilles-peskine-arm committed:
Update references to headers that have been renamed

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

gilles-peskine-arm committed:
Minor comment fixes

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

gilles-peskine-arm committed:
Fix auto-enabling of MD for PKCS5

This was done too late, so it could have led to mysterious build errors if
PKCS5 was the only thing that requires MD.

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

gilles-peskine-arm committed:
Remove enforcement against direct inclusion of private headers

We added `#error` directives to prevent some headers from being included
directly because those headers were not clearly indicated as public (and, in
the case of `check_config.h`, were formerly public). Headers under
`.../private` have no history of being public and are obviously not intended
for direct consumption, so remove the enforcement.

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

ronald-cron-arm commented:
One minor comment remaining. Otherwise it looks good to me.

gilles-peskine-arm committed:
Update comments about MBEDTLS_PLATFORM_SNPRINTF_ALT on Windows

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

ronald-cron-arm reviewed (dismissed):
LGTM, thanks.

minosgalanakis reviewed (dismissed):
LGTM

gilles-peskine-arm commented:
I rebased to include the merge of the framework prerequisite and to resolve the framework conflict. The previous version is in https://github.com/gilles-peskine-arm/TF-PSA-Crypto/tree/build_info-cleanup-1.0-4

ronald-cron-arm reviewed (approved):
Rebase and framework update look good to me.

minosgalanakis reviewed (approved):
LGTM
